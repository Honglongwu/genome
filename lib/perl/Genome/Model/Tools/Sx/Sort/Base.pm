package Genome::Model::Tools::Sx::Sort::Base;

use strict;
use warnings;

use Genome;

require IPC::Open2;

class Genome::Model::Tools::Sx::Sort::Base {
    is  => 'Genome::Model::Tools::Sx::Base',
};

sub help_brief {
    return 'Sort sequences by average quality';
}

sub execute {
    my $self = shift;

    my $init = $self->_init;
    return if not $init;

    my $pid = IPC::Open2::open2(\*SORTED, \*UNSORTED, 'sort', $self->_sort_params);
    if ( not $pid ) {
        $self->error_message('Failed to create the sort command via open2');
        return;
    }

    my $flatten = $self->can('_flatten');
    my $inflate = $self->can('_inflate');

    my $reader = $self->_reader;
    while ( my $seqs = $reader->read ) {
        print UNSORTED ($flatten->(@$seqs), "\n");
    }
    close UNSORTED;

    my $writer = $self->_writer;
    while ( my $line = <SORTED> ) {
        $writer->write( $inflate->($line) );
    }
    close SORTED;
    waitpid($pid, 0);

    return 1;
}

1;

