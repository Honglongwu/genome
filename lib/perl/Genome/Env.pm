package Genome::Env;

use strict;
use warnings;

use Module::Find qw(findsubmod usesub);
use Set::Scalar qw();

sub used {
    return grep { /^GENOME_/ } keys %ENV;
}

sub allowed_modules {
    return usesub Genome::Env;
}

sub allowed {
    return map { NAME($_) } allowed_modules();
}

sub check_genome_variables {
    my @valid = allowed();
    my @used  = used();

    my $valid_set = Set::Scalar->new(@valid);
    my $used_set = Set::Scalar->new(@used);
    my $invalid_set = $used_set - $valid_set;

    unless ($invalid_set->is_empty) {
        print STDERR "Available environment variable(s):\n",
            '  ', join("\n  ", @valid), "\n\n";
        print STDERR "Unrecognized environment variable(s) found:\n",
            '  ', join("\n  ", map { join('=', $_, ($ENV{$_} || '')) } $invalid_set->members), "\n";
        return;
    }

    return 1;
}

sub set_default_values {
    my @valid = allowed_modules();
    for my $module (@valid) {
        my $name = NAME($module);
        if ($module->can('default_value')
            && !defined($ENV{$name})
        ) {
            $ENV{$name} = $module->default_value;
        }
    }
}

sub NAME {
    my $class = shift;

    my $tail = ($class =~ /^Genome::Env::(.*)/)[0];
    my $NAME = uc($tail);
    unless ($NAME) {
        die 'Could not infer NAME from module';
    }
    return $NAME;
}

# Load all modules in Genome/Env/* and set their default value
# unless the environment variable is already defined.
set_default_values();

# Make sure that all Genome environment variables that are
# set correspond to a file in Genome/Env/*.
check_genome_variables(); # instead of 1; this should be last statment
