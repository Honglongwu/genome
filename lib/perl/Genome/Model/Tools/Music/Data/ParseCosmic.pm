package Genome::Model::Tools::Music::Data::ParseCosmic;

use warnings;
use strict;
use IO::File;

our $VERSION = $Genome::Model::Tools::Music::VERSION;

class Genome::Model::Tools::Music::Data::ParseCosmic {
    is => 'Command::V2',
    has_input => [
        input_complete_file => { is => 'String', doc => "CosmicCompleteExport downloaded from cosmic" },
        input_ins_mut_file => { is => 'String', doc => "CosmicInsMutExport downloaded from cosmic" },
        output_file => { is => 'String', doc => "Cleaned up and standardized variant list with sample IDs" },
    ],
    has_optional_input => [
        hg18_fasta => { is => 'String', doc => "Reference FASTA file for HG18 (Build36)",
                        example_values => ["/gscmnt/gc4096/info/model_data/2741951221/build101947881/all_sequences.fa"] },
        hg19_fasta => { is => 'String', doc => "Reference FASTA file for HG19 (Build37)",
                        example_values => ["/gscmnt/ams1102/info/model_data/2869585698/build106942997/all_sequences.fa"] },
        max_indel_length => { is => 'Integer', doc => "Skip indels longer than this many bps", default => 100 },
        genome_wide_only => { is => 'Boolean', doc => "Pull only variants from genome-wide or exome-wide screens", default => 0 },
    ],
    doc => "Parses and standardizes nucleotide-level variants downloaded from COSMIC",
};

sub help_detail {
    return <<HELP;
The two expected input files are downloadable from Sanger's Public FTP at:
ftp://ftp.sanger.ac.uk/pub/CGP/cosmic/data_export

"input_complete_file" should be the complete mutation list from COSMIC:
CosmicCompleteExport_v62_291112.tsv

"input_ins_mut_file" should be the separate file with nucleotide sequences for somatic insertions:
CosmicInsMutExport_v62_291112.tsv

Shown above are the filenames from COSMIC release v62. You can find other versions of these files
on Sanger's FTP. This tool should work fine with most of them.

Note that this tool will skip the following based on data available in COSMIC:

=over

 - variants that don't have both nucleotide change and genomic loci available
 - variants with reference alleles that don't match the provided hg18/hg19 reference sequence
 - variants annotated as germline, or from samples annotated as normals
 - variants annotated as belonging to metastases, adjacent tissue, recurrent tumors, etc.
 - insertions whose sequence don't match the separate cosmic file for somatic insertions
 - deletions whose length don't match the genomic loci
 - TNPs, ONPs and complex indels (DNPs are not skipped)
HELP
}

#########################################################################

sub execute {
    my $self = shift;
    my $input_complete_file = $self->input_complete_file;
    my $input_ins_mut_file = $self->input_ins_mut_file;
    my $hg18_fasta = $self->hg18_fasta;
    my $hg19_fasta = $self->hg19_fasta;
    my $max_indel_length = $self->max_indel_length;
    my $genome_wide_only = $self->genome_wide_only;
    my $output_file = $self->output_file;

    # Some hashes for quick lookup
    my %valid_chrom = map {($_,1)} ( 1..22, qw( X Y MT ));
    my %complement = qw( A T T A C G G C N N );

    # Load up the sequences of insertions, which COSMIC happens to store separately for whatever reason
    my %ins_nucs = map{chomp; split(/\t/)} grep {m/^\w+\t[ACGTacgt]+$/} `cut -f 9,11 $input_ins_mut_file`;

    # From the main cosmic file, parse out genomic loci and ref/var for SNVs and indels
    my %vars = ();
    my %tissue_site_counts = ();
    my $cosmicFh = IO::File->new( $input_complete_file );
    warn "Parsing COSMIC DB for usable SNVs and indels...\n";
    while( my $line = $cosmicFh->getline ) {
        next if( $line =~ m/^Gene name/ );
        chomp( $line );
        my @col = split( /\t/, $line );
        my ( $samp_name, $samp_id, $samp_site, $samp_hist, $samp_subtype, $wgs, $origin ) = @col[3,4,6,8,9,10,23];
        my ( $mut_id, $mut_nuc, $mut_aa, $mut_type, $zygos, $hg18_locus, $hg18_strand, $hg19_locus, $hg19_strand, $status ) = @col[11..20];

        # Skip variants of samples that were not "Genome-wide screened" (if the flag is turned on)
        next if( $genome_wide_only and $wgs ne 'y' );

        # Skip variants belonging to recurrent tumors, metastases, adjacent, etc. - We don't want to report the same variant from the same patient more than once
        next unless( $origin =~ m/^(primary|secondary|surgery)/ );

        # Skip known germline variants or variants from samples annotated as normals
        next if( $status =~ m/germline/ or $samp_hist =~ m/NORMAL/ or $samp_subtype =~ m/normal/);

        # Skip mutation types that we cannot deal with
        next if( $mut_type =~ m/^Whole gene deletion/ );

        # Skip variants that don't have both nucleotide change and loci available
        next unless( $mut_nuc and ( $hg18_locus or $hg19_locus ));

        # Pull b36 or b37 loci of the variant giving preference to b36, which we will liftOver later
        my $build = ( $hg18_locus ? "b36" : "b37" );
        my $locus = ( $hg18_locus ? $hg18_locus : $hg19_locus );
        my ( $chr, $start, $stop ) = $locus =~ m/^(\w+):(\d+)-(\d+)$/;
        die "Cannot identify locus for:\n$line\n" unless( $chr and $start and $stop );
        $chr =~ s/^(23|x)$/X/; $chr =~ s/^(24|y)$/Y/; $chr =~ s/^(25|M|m|mt)$/MT/;
        die "Invalid chrom name in:\n$line\n" unless( $valid_chrom{$chr} );

        # Fetch the reference sequence at this locus to do some QC
        my $ref_seq = ( $hg18_locus ? $hg18_fasta : $hg19_fasta );
        my $decr_start = $start - 1;
        my $fetched_ref = `echo "$chr\t$decr_start\t$stop" | joinx ref-stats --ref-bases - $ref_seq | grep -v ^# | cut -f 7`;
        chomp( $fetched_ref );

        # Try to find out what kind of variant this is, and pull ref/var if possible
        my ( $ref, $var, @tmp );
        if( @tmp = $mut_nuc =~ m/^c\.\d+[+-]*\d*_*\d*[+-]*\d*([ACGTacgt]+)>([ACGTacgt]+)$/ ) { # SNV
            ( $ref, $var ) = ( uc( $tmp[0] ), uc( $tmp[1] ));
        }
        elsif( @tmp = $mut_nuc =~ m/^c\.\d+[+-]*\d*_*\d*[+-]*\d*ins([ACGTacgt]+)$/ ) { # insertion
            ( $ref, $var ) = ( "-", uc( $tmp[0] ));
            my $ins_nuc = uc( $ins_nucs{$mut_id} );
            if( length( $var ) > $max_indel_length ) {
                warn "Skipped: Long insertion >$max_indel_length bps in $mut_nuc at $build locus: $locus\n";
                next;
            }
            elsif( $ins_nuc and $var ne $ins_nuc ) {
                warn "Skipped: Inserted sequence per COSMIC ($ins_nuc) differs from $mut_nuc at $build locus: $locus\n";
                next;
            }
        }
        elsif( @tmp = $mut_nuc =~ m/^c\.\d+[+-]*\d*_*\d*[+-]*\d*del([ACGTacgt]+)$/ ) { # deletion
            ( $ref, $var ) = ( uc( $tmp[0] ), "-" );
            if( length( $ref ) > $max_indel_length ) {
                warn "Skipped: Long deletion >$max_indel_length bps in $mut_nuc at $build locus: $locus\n";
                next;
            }
            elsif( length( $ref ) != ( $stop - $start + 1 )) {
                warn "Skipped: Length of deleted sequence in $mut_nuc doesn't match $build locus: $locus\n";
                next;
            }
        }
        elsif( @tmp = $mut_nuc =~ m/^c\.\d+[+-]*\d*_*\d*[+-]*\d*del(\d+)$/ ) { # deletion without sequence specified
            my $del_length = $tmp[0];
            if( $del_length > $max_indel_length ) {
                warn "Skipped: Long deletion >$max_indel_length bps in $mut_nuc at $build locus: $locus\n";
                next;
            }

            if( length( $fetched_ref ) != $del_length ) {
                warn "Skipped: Length of deleted sequence in $mut_nuc doesn't match $build locus: $locus\n";
                next;
            }
            ( $ref, $var ) = ( uc( $fetched_ref ), "-" );
        }

        unless( $ref and $var ) {
            warn "Skipped: Cannot identify ref/var from $mut_nuc at $build locus: $locus\n";
            next;
        }

        unless(( length($ref)==1 && length($var)==1 ) || ( length($ref)==2 && length($var)==2 ) ||
               ( $ref eq "-" && $var=~m/[ACGT]+/ ) || ( $ref=~m/[ACGT]+/ && $var eq "-" )) {
            warn "Skipped: Cannot handle $ref/$var at $build locus: $locus\n";
            next;
        }

        # Reverse complement the ref/var if the variant is from the negative strand
        my $locus_strand = ( $hg18_strand ? $hg18_strand : $hg19_strand );
        my $rc_ref = reverse map {$complement{$_}} split( //, $ref );
        if( $locus_strand eq "-" and $rc_ref eq $fetched_ref ) {
            if( $ref eq "-" && $var =~ m/[ACGT]+/ ) {
                $var = reverse map {$complement{$_}} split( //, $var );
            }
            elsif( $ref =~ m/[ACGT]+/ && $var eq "-" ) {
                $ref = $rc_ref;
            }
            else {
                $ref = $rc_ref;
                $var = reverse map {$complement{$_}} split( //, $var );
            }
        }

        # Skip SNVs and deletions with reference alleles that don't match the reference fasta
        if( $ref ne $fetched_ref and $ref ne "-" ) {
            warn "Skipped: Variant's ref allele $ref is $fetched_ref in the $build fasta at locus: $locus\n";
            next;
        }

        # If we get to this point, then we have all the information we need. Save it into a hash
        $samp_site = $samp_hist if( $samp_site eq "NS" );
        ++$tissue_site_counts{$samp_site} unless( defined $vars{$samp_id} );
        $vars{$samp_id}{$build}{"$chr\t$start\t$stop\t$ref\t$var"} = 1;
    }
    $cosmicFh->close;

    # Remove output file if it already exists
    unlink $output_file if( -e $output_file );

    # De-deduplicate variants by sample ID, and dump it into the same output file
    my ( $samp_count, $tot_mut_count ) = ( 0, 0 );
    warn "\nRunning liftOver and prepping Build37 files in a more familiar format...\n";
    foreach my $samp_id ( keys %vars ) {

        # Make 2 temporary files to use with liftover
        my $build36_file = Genome::Sys->create_temp_file_path;
        my $build37_file = Genome::Sys->create_temp_file_path;
        ( $build36_file and $build37_file ) or die "Couldn't create a temp file. $!";

        # Write any build36 variants into a 5 column WU format for use with liftOver
        my $mut_count = 0;
        if( defined $vars{$samp_id}{b36} && scalar( keys %{$vars{$samp_id}{b36}} ) > 0 ) {
            $mut_count += scalar( keys %{$vars{$samp_id}{b36}} );
            my $outFh = IO::File->new( $build36_file, ">" );
            foreach my $line ( keys %{$vars{$samp_id}{b36}} ) {
                $outFh->print( "$line\t$samp_id\n" );
            }
            $outFh->close;

            # Run liftOver using the GMT wrapper (::TODO:: use liftOver binary instead)
            my $lift_cmd = Genome::Model::Tools::LiftOver->create(
                source_file => $build36_file, destination_file => $build37_file,
                input_is_annoformat => 1, lift_direction => "hg18ToHg19",
            );
            ( $lift_cmd->execute ) or die "Failed to run 'gmt lift-over' for variants in $samp_id\n";
            $lift_cmd->delete;
        }

        # Append additional b37 loci to the liftOver'ed file, and remove any duplicates
        if( defined $vars{$samp_id}{b37} && scalar( keys %{$vars{$samp_id}{b37}} ) > 0 ) {
            $mut_count += scalar( keys %{$vars{$samp_id}{b37}} );
            my $outFh = IO::File->new( $build37_file, ">>" );
            foreach my $line ( keys %{$vars{$samp_id}{b37}} ) {
                $outFh->print( "$line\t$samp_id\n" );
            }
            $outFh->close;
        }
        print `sort -u $build37_file -o $build37_file`;

        # Sort by loci, and write to output file
        if( -s $build37_file ) {
            ++$samp_count;
            $tot_mut_count += $mut_count;
            print `joinx sort -s -i $build37_file >> $output_file`;
        }
    }

    print "\nFetched $tot_mut_count SNVs and indels across $samp_count samples from the following tissue types:\n";
    foreach my $key ( sort {$tissue_site_counts{$b} <=> $tissue_site_counts{$a}} keys %tissue_site_counts ) {
        print $tissue_site_counts{$key} . "\t$key\n";
    }

    return 1;
}

1;
