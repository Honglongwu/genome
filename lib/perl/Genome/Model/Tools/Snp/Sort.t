#!/usr/bin/env genome-perl

use strict;
use warnings;

use Test::More tests => 4;
use File::Compare;

use above 'Genome';

BEGIN {
        use_ok('Genome::Model::Tools::Snp::Sort');
    };

my $test_snp_file = $ENV{GENOME_TEST_INPUTS} . '/Genome-Model-Tools-Snp/Sort/test.snp';
my $expected_out_file = $ENV{GENOME_TEST_INPUTS} . '/Genome-Model-Tools-Snp/Sort/test.out';

my $tmp_dir =  File::Temp::tempdir('gt-snp-sort-XXXXXX',DIR => "$ENV{GENOME_TEST_TEMP}",CLEANUP => 1);
my $output_file = $tmp_dir .'/test.out';
my $snp_sorter = Genome::Model::Tools::Snp::Sort->create(
                                                         snp_file => $test_snp_file,
                                                         output_file => $output_file,
                                                     );
isa_ok($snp_sorter,'Genome::Model::Tools::Snp::Sort');
ok($snp_sorter->execute,'execute command '. $snp_sorter->command_name);
ok(!compare($expected_out_file,$output_file),'sorted output file matches expected output');

exit;
