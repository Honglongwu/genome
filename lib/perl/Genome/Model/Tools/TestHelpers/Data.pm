package Genome::Model::Tools::TestHelpers::Data;

use strict;
use warnings;

use Params::Validate qw(:types);
use above 'Genome';
use Genome::Model::Tools::TestHelpers::General qw(
    get_test_data
);
use Test::More;

require Exporter;
our @ISA = qw(Exporter);
our @EXPORT_OK = qw(
    get_shared_test_data
);

sub get_shared_test_data {
    my ($data_path, $version) = Params::Validate::validate_pos(@_,
        1, 1);

    return get_test_data('Genome::Model::Tools::TestHelpers::Data',
        $data_path, $version);
}

1;
