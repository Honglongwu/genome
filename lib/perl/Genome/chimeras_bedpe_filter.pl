#!/usr/bin/env perl
#use warnings;
use strict;
#use diagnostics;
use File::Basename qw(basename);

#### Include get options and usage
#
# (1) common 5' and 3' partners
# (2) 
#

######################################################################
#                       Process ChimeraScan Results                  #
######################################################################
#my $dir = '/gscmnt/gc5105/research/cmaher/LungCancer/WIS/CSCAN/RESULTS/';
#my $dir = '/gscmnt/gc5105/research/cmaher/LungCancer/17genomes/CSCAN/RESULTS/';
#my $dir = '/gscmnt/gc6127/research/cmaher/4Malachi/RAW/';
#my $dir = '/gscmnt/gc6127/research/cmaher/Myeloma/MMY5/';
#my $dir = '/gscmnt/gc6127/research/cmaher/ALL/ALL1/';
my $dir = '/gscmnt/gc7001/info/model_data/2892653587/build139185824/fusions/';

#my $outfile = '/gscmnt/gc5105/research/cmaher/LungCancer/17genomes/LUC9/cnv.bed';
#open(OUTFILE, ">$outfile");

# Mitelman database
my $mitelman = '/gscuser/cmaher/References/molclingene.dat';
my (%MITEL,%MITEL5P,%MITEL3P);
open(MITEL, "<$mitelman" ) or die("Couldn't open MITELMAN file $mitelman \n");
while(<MITEL>)  { 
    chomp;
    my(@v)=split(/\t/);
    
    if($v[5] =~ /\//){
	$MITEL{$v[5]}[0]++; # Raise SNV Sample Count
    }
}
close(MITEL);

my @fusions = keys %MITEL;
foreach my $fusion (@fusions){
    my($FP,$TP) = split(/\//,$fusion);
    $MITEL5P{$FP}[0]++;
    $MITEL3P{$TP}[0]++;
}

my @FPgenes = keys %MITEL5P;
my @TPgenes = keys %MITEL3P;

# Drug Targets



# Kinases
my $kinase = '/gscmnt/gc5105/research/cmaher/Reference/Phosphatases_and_Kinases.txt';
my (%KIN);
open(K, "<$kinase" ) or die("Couldn't open K file $kinase \n");
while(<K>)  { 
    chomp;
    my(@v)=split(/\t/);
    
    if($v[1] ne ''){
	$KIN{$v[1]}[0]++; # Raise SNV Sample Count
    }
}
close(K);


# Cancer Genes
my $cancer = '/gscmnt/gc5105/research/cmaher/Reference/CancerGenes.1172012.txt';
my %CANCER;
open(C,"<$cancer") or die("Couldn't open C file $cancer \n");
while(<C>){
    chomp;
    my(@v)=split(/\t/);
    
    $CANCER{$v[0]}[0]++;
}
close(C);






my (%GF,%FPR,%TPR);
my @samples;
#
# Extract Recurrent
#
my $x = '5'; # Element in array
#my $dir = '/gscmnt/gc5105/research/cmaher/LungCancer/17genomes/CSCAN/RESULTS/';
foreach my $gffile (glob $dir."*chimeras.bedpe"){
    my($sample,$c,$file_ext) = split(/\./,basename $gffile);
    push @samples, $sample;
    
    my(%FPPROM,%TPPROM);
    open(GF, "<$gffile" ) or die("Couldn't open GF file file $gffile \n");
    while(<GF>)  { 
	chomp;
	
	if(/^\#/){ # Bypass headers
	    next;
	}else{
	    my(@v)=split(/\t/);
	    my $FP = $v[12];
	    my $TP = $v[13];
	    $FPPROM{$FP}[0]++;
	    $TPPROM{$TP}[0]++;
	}
    }
    close(GF);

    #print "Processing $sample $gffile ...\n";
    open(GF, "<$gffile" ) or die("Couldn't open GF file file $gffile \n");
    while(<GF>)  { 
	chomp;
	
	if(/^\#/){ # Bypass headers
	    next;
	}else{
	    my(@v)=split(/\t/);
	    my $FP = $v[12];
	    my $TP = $v[13];
	    
	    my $score = $v[7];
	    my $total_frag = $v[16];
	    my $span_frag = $v[17];
	    my $type = $v[14]; 
	    
	    if($FP eq $TP){ next; }
	    
	    my $fusion = $FP.':'.$TP;

	    
	    if($v[21] =~ /AAAAAAAAAA/ || $v[21] =~ /TTTTTTTTTT/ || $v[21] =~ /GGGGGGGGGG/ || $v[21] =~ /CCCCCCCCCC/){
		next;
	    }
	 

	    #print "$sample\t$fusion\t$FPPROM{$FP}[0]\t$TPPROM{$TP}[0]\n";
	    if($FPPROM{$FP}[0] <= 3 && $TPPROM{$TP}[0] <= 3){ #filter out fusions where either gene is involved in many different fusions
		#if($fusion =~ /CREB3/){ print "PASS\t$fusion\n"; }
		if($total_frag ne $span_frag){
#		if( ($total_frag > '30' && $total_frag ne $span_frag) || $total_frag <= '30'){
		    $GF{$fusion}[0]++;
		    $GF{$fusion}[1]+=$total_frag;
		    $GF{$fusion}[2]+=$span_frag;
		    $GF{$fusion}[3]=$type;
		    $GF{$fusion}[4]+=$score;
		    $GF{$fusion}[$x] = $span_frag.':'.$total_frag;
		    #$FPR{$FP}[0]++;
		    #$TPR{$TP}[0]++;
		}
	    }
	}
    }
    close(GF);
    $x++;
}

#close(OUTFILE);

# Print header
print "Fusion\t5P\t3P\tTotal_Freq\tSpanning_Freq\tType\tScore\tSpan:Total\tMitel5P\tMitel3P\tKinase5P\tKinase3P\tCancer5P\tCancer3P\n"; #5'PartnersFreq\t3'PartnersFreq";
for(my $x='0';$x<@samples;$x++){ print "\t$samples[$x]"; }
print "\n";

# Print output
my @fusions = keys %GF;
for(my $i = '0'; $i < @fusions; $i++){
    
    # Get count
    my $sample_freq = '0';
    my $total_freq = '0';
    for(my $j = '5'; $j < @samples + 5; $j++){ # Only check data columns
	if($GF{$fusions[$i]}[$j] ne ''){
	    $total_freq++;
	    my($s,$e)=split(/\:/,$GF{$fusions[$i]}[$j]);
	    
	    if($s ne '0'){
#		if($GF{$fusions[$i]}[1] >= '3' && $GF{$fusions[$i]}[2] >= '1'){
		    $sample_freq++;
#		}
	    }
	}
    }
    
    #if($sample_freq eq '1' && $total_freq eq '1' && $GF{$fusions[$i]}[1] >= '2' && ($GF{$fusions[$i]}[1] ne $GF{$fusions[$i]}[2]) ){
    if( ($sample_freq eq '1' || $sample_freq eq '2') && $GF{$fusions[$i]}[1] >= '5' && $GF{$fusions[$i]}[2] >= '1'){
	                                                                                                                                   
	my($gene1,$gene2)=split(/\:/,$fusions[$i]); 
	print "$fusions[$i]\t$gene1\t$gene2";

	# Print all
	for(my $j = '1'; $j < @samples + 5; $j++){ # Exclude first column and instead print actual sample frequency
	    if($GF{$fusions[$i]}[$j] eq ''){
		print "\t0";
	    }else{
		my($s,$e)=split(/\:/,$GF{$fusions[$i]}[$j]);
		print "\t$GF{$fusions[$i]}[$j]";
	    }
	}

#$sample_freq\t$total_freq";
       
	if($MITEL5P{$gene1}[0] ne ''){
	    print "\tMITEL5_".$gene1;
	}else{
	    print "\tNull";
	}
	
	if($MITEL3P{$gene2}[0] ne ''){
	    print "\tMITEL3_".$gene2;
	}else{ print "\tNull"; }
	
	if($KIN{$gene1}[0] ne ''){
	    print "\tKIN5_".$gene1;
	}else{ print "\tNull"; }
	
	if($KIN{$gene2}[0] ne ''){
	    print "\tKIN3_".$gene2;
	}else{ print "\tNull"; }
	
	if($CANCER{$gene1}[0] ne ''){
	    print "\tCAN5_".$gene1;
	}else{ print "\tNull"; }
	
	if($CANCER{$gene2}[0] ne ''){
	    print "\tCAN3_".$gene2;   
	}else{ print "\tNull"; }
	
	if($FPR{$gene1}[0] > '1'){
	    print "\tFPR_".$FPR{$gene1}[0];
	}else{ print "\tNull"; }
	
	if($TPR{$gene2}[0] > '1'){
	    print "\tTPR_".$TPR{$gene2}[0];
	}else{ print "\tNull"; } 
	
	print "\n";
	
    }
}
