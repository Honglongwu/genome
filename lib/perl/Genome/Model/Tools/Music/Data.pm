package Genome::Model::Tools::Music::Data;
use warnings;
use strict;
use Genome;

our $VERSION = $Genome::Model::Tools::Music::VERSION; 

class Genome::Model::Tools::Music::Data {
    is => ['Command::Tree'],
    doc => "Parse, standardize, and index various third-party datasets used by the tools in MuSiC."
};

sub _doc_copyright_years {
    (2010,2013);
}

sub _doc_license {
    my $self = shift;
    my (@y) = $self->_doc_copyright_years;  
    return <<EOS
Copyright (C) $y[0]-$y[1] Washington University in St. Louis.

It is released under the Lesser GNU Public License (LGPL) version 3.  See the 
associated LICENSE file in this distribution.
EOS
}

sub _doc_authors {
    return <<EOS
 Cyriac Kandoth, Ph.D.
EOS
}

=cut
sub _doc_credits {
    # used to compose man pages
    # (return a list of strings)
    return ('','None at this time.');
}
=cut

sub _doc_see_also {
    return <<EOS
B<genome-music>(1),
B<genome>(1)
EOS
}

sub _doc_manual_body {
    return shift->help_detail;
}

sub help_detail {
    return "These tools are part of the MuSiC suite.\n";
}

1;
