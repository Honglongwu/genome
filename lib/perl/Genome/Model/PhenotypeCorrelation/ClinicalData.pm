package Genome::Model::PhenotypeCorrelation::ClinicalData;

use Digest::MD5;
use Genome;
use Sort::Naturally qw/nsort ncmp/;
use Data::Dumper;
use Carp qw/confess/;

use strict;
use warnings;

my $CATEGORICAL_THRESHOLD = 0.05;

class Genome::Model::PhenotypeCorrelation::ClinicalData {
    is => "UR::Object",
    doc => "Clinical (phenotype) data for a set of samples",
    has_transient_optional => [
        _attributes => {
            is => "HASH",
            doc => "Contains type information (datatype, quantitative vs categorical, etc.)",
        },
        _sample_indices => {
            is => "HASH",
        },
        _phenotypes => {
            is => "HASH",
        },
    ],
};

sub sample_count {
    my $self = shift;
    return scalar keys %{$self->_sample_indices};
}

sub attributes {
    my ($self, %params) = @_;
    if (defined $params{categorical}) {
        return nsort grep {
                $self->_attributes->{$_}->{categorical} == $params{categorical}
            } keys %{$self->_attributes};
    }
    return nsort keys %{$self->_attributes};
}

sub attribute_types {
    my $self = shift;
    return $self->_attributes;
}

sub _validate_attribute {
    my ($self, $attr) = @_;
    unless (defined $self->_phenotypes->{$attr}) {
        my $valid_attrs = join("\n\t", $self->attributes);
        confess "Unknown attribute name '$attr'. Valid attributes are\n\t$valid_attrs";
    }
}

sub attribute_values {
    my ($self, $attr) = @_;
    $self->_validate_attribute($attr);
    return $self->_phenotypes->{$attr};
}

sub attribute_value_for_sample {
    my ($self, $sample_name, $attr) = @_;
    $self->_validate_attribute($attr);
    my $sample_idx = $self->_sample_indices->{$sample_name};
    return $self->_phenotypes->{$attr}->[$sample_idx];
}

sub from_database {
    my ($class, $nomenclature, @samples) = @_;
    my $self = $class->create;

    @samples = sort { ncmp($a->name, $b->name) } @samples;
    my %sample_indices;
    @sample_indices{map {$_->name} @samples} = 0..$#samples;
    $self->_sample_indices(\%sample_indices);

    $DB::single=1;
    my %phenotypes;
    my %attributes;
    foreach my $sample (@samples) {
        my @attrs = $sample->source->attributes_for_nomenclature($nomenclature);
        my $sample_idx = $sample_indices{$sample->name};

        foreach my $attr (@attrs) {
            my $label = $attr->attribute_label;
            $attributes{$label} = {};

            $phenotypes{$label} = [(undef) x scalar @samples] unless defined $phenotypes{$label};
            $phenotypes{$label}->[$sample_idx] = $attr->attribute_value;
        }
    }
    $self->_phenotypes(\%phenotypes);

    confess "No clinical attributes found!" unless scalar(keys %attributes);

    for my $attr (keys %attributes) {
        $attributes{$attr}->{categorical} = $self->_is_categorical($attr);
    }
    $self->_attributes(\%attributes);

    return $self;
}

sub from_file {
    my ($class, $fh, %params) = @_;
    my $self = $class->create;
    my $missing_string = $params{missing_string} || "NA";

    my $header_line = $fh->getline or confess "Failed to read clinical data header";
    chomp $header_line;
    my @attr_names = split("\t", $header_line);
    shift @attr_names; # drop the leading Sample_name in header

    my @sample_data;
    while (my $line = $fh->getline) {
        chomp $line;
        my @fields = map { $_ eq $missing_string ? undef : $_ } split("\t", $line);
        push(@sample_data, \@fields);
    }

    @sample_data = sort { ncmp($a->[0], $b->[0]) } @sample_data;
    my %sample_indices;
    @sample_indices{map {$_->[0]} @sample_data} = 0..$#sample_data;

    my %phenotypes;
    @phenotypes{@attr_names} = map { [(undef) x scalar(@sample_data)] } @attr_names;
    for my $i (0..$#sample_data) {
        my @values = @{$sample_data[$i]};
        my $sample_name = shift @values;
        for my $j (0..$#attr_names) {
            $phenotypes{$attr_names[$j]}->[$i] = $values[$j];
        }
    }

    $self->_sample_indices(\%sample_indices);
    $self->_phenotypes(\%phenotypes);
    my %attributes = map { $_ => { categorical => $self->_is_categorical($_) } } @attr_names;
    $self->_attributes(\%attributes);

    return $self;
}

sub write_file {
    my ($self, $fh, %params) = @_;
    my $missing_string = $params{missing_string} || "NA";
    my @attr_names;
    if (defined $params{attribute_names}) {
        @attr_names = @{$params{attribute_names}};
        for my $attr (@attr_names) {
            confess "unknown clinical attribute $attr" unless defined $self->_attributes->{$attr};
        }
    } else {
        @attr_names = nsort keys %{$self->_attributes};
    }

    my $md5 = new Digest::MD5;
    my $write = sub {
        my $data = shift;
        $md5->add($data);
        $fh->write($data);
    };

    &$write("Sample_name\t" . join("\t", @attr_names) . "\n");
    for my $sample_name (nsort keys %{$self->_sample_indices}) {
        my $sample_idx = $self->_sample_indices->{$sample_name};
        my @values = map {
             defined $self->_phenotypes->{$_}->[$sample_idx] ?
                $self->_phenotypes->{$_}->[$sample_idx] : $missing_string
            } @attr_names;
        &$write("$sample_name\t" . join("\t", @values) . "\n");
    }

    return $md5->hexdigest;
}

sub _is_categorical {
    my ($self, $attribute) = @_;
    my $values = $self->_phenotypes->{$attribute};
    confess "Unknown attribute '$attribute'" unless defined $values;
    my %uniq = map {$_ => 1} grep {$_} @$values;
    my $n_uniq = scalar keys(%uniq);
    my $ratio = $n_uniq / $self->sample_count;
    my $decision = ($n_uniq <= 10) || ($ratio <= $CATEGORICAL_THRESHOLD);
    return $decision;
}


1;
