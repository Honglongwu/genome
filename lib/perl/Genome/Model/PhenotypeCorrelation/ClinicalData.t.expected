../../../../../test-data-relocate-me/lib--perl--Genome--Model--PhenotypeCorrelation--ClinicalData.t.expected